### PR TITLE
Add the cardinality check to detect ambiguous target row for MERGE INTO

### PR DESCRIPTION
**Pull request by dilipbiswal**
SQL standard requires an exception to be thrown if the ON clause in MERGE is such that more than 1 row in source matches a row in target

Implements suggestion 1 from Anton in [comment](https://github.com/apache/iceberg/pull/1947#issuecomment-747450897)

```

1. append _row_id and _file columns to the target table
2. do an inner join on the merge condition
3. define an udf that accepts the file name, adds it to the accumulator and retuns 1 => Changed it to an expression to stay in catalyst world.
4. group by _row_id, perform the cardinality check
5. access the accumulator to get the matching files

```

### CODE REVIEW

**Comment by dilipbiswal** [comment]
@aokolnychyi @rdblue 
Kindly review the top commit which has the cardinality check implementation.

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 1]
```diff
@@ -164,4 +164,7 @@ private TableProperties() {
 
   public static final String MERGE_MODE = "write.merge.mode";
   public static final String MERGE_MODE_DEFAULT = "copy-on-write";
+
+  public static final String MERGE_WRITE_CARDINALITY_CHECK = "write.merge.cardinality.check.enabled";
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:168]: I like Anton's original suggestion better than this property. How about `write.merge.cardinality-check.enabled`?

**Comment by rdblue** [comment]
Sorry I didn't look into this more thoroughly before now, but I think that there is a better way to implement the cardinality check that doesn't require an inner join.

The (2003) SQL spec in Part 2, section 14.9 says:

> Let Q be the result of evaluating [the USING clause]
> If <merge when matched clause> is specified, then:
> For each row R1 of T:
> ...
> Let M be the number of matching rows in Q for R1.
> If M is greater than 1 (one), then an exception condition is raised: cardinality violation.

I think that means the cardinality check is done for each row of the target table. If two rows from the target table have the same "key" from the `ON` condition, that's okay and they are processed separately. The cardinality violation applies when there are two rows from the result of the `USING` clause for any row.

That aligns with [a comment from Eugene Koifman](https://community.cloudera.com/t5/Support-Questions/Hive-Merge-command-throwing-error-message/td-p/228500) (from the Hive implementation):

> Logically what this means is that the query is asking the system to update 1 existing row in target in 2 (or more) different ways.

As a result, I think that there is an easier way to implement this check. Instead of running an additional inner join, the check should run as the first thing in [`processRow`](https://github.com/apache/iceberg/pull/2022/files#diff-deb1276df77b1ac20d203d4607231c593cfeac1a3c3771d174bd04fc1afd773aR92):
1. Add `monotonicaly_increasing_id` to the target table rows
2. In `processRow`, keep track of the last row's ID
3. If the last row's ID matches the current row's ID, throw an exception for the cardinality check
4. If the current row's ID is different, then replace the last row ID with it

That works because the join will always produce the matching rows together as a group.

**Comment by dilipbiswal** [comment]
@rdblue Thanks a lot. I am just thinking if 
```
That works because the join will always produce the matching rows together as a group.
```
is true in all cases. @aokolnychyi What do you think ?

Ryan, one other thing is , we are not introducing an additional join for the cardinality check in this implementation. We are doing the same join as we are doing before but just accumulating the files using an accumulator.

**Comment by rdblue** [comment]
I think it is true in all cases because all of the matches for a target row must be produced when that target row is considered.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala [hunk 2]
```diff
@@ -77,7 +61,7 @@ object RewriteDelete extends Rule[LogicalPlan] with PredicateHelper with Logging
       val writeInfo = newWriteInfo(r.schema)
       val mergeBuilder = r.table.asMergeable.newMergeBuilder("delete", writeInfo)
 
-      val scanPlan = buildScanPlan(r.table, r.output, mergeBuilder, cond)
+      val scanPlan = buildScanPlan(spark, r.table, r.output, mergeBuilder, Some(cond))
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala:64]: I would prefer to keep changes for MERGE INTO separate from DELETE FROM right now. That will keep PRs smaller and we can focus on getting MERGE INTO right, without the distraction of sometimes breaking DELETE FROM. Once we are happy with it, then we can refactor to produce common utilities.

Comment (dilipbiswal) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala:64]: @rdblue Okay Ryan.

**Comment by dilipbiswal** [comment]
Had a offline discussion with Ryan. Since we can not guarantee rids to be emitted in sequence without some additional work (yet to be determined), we will take a conservative approach in the first pass.

cc @aokolnychyi @rdblue

**Comment by aokolnychyi** [comment]
Thanks for the PR, @dilipbiswal.

Sorry, I got distracted by other things. I'll review this first thing in the morning tomorrow.

**Comment by aokolnychyi** [comment]
Did I get correctly that we cannot guarantee the rows will be emitted in the expected sequence, @dilipbiswal @rdblue?

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                    performCountCheckForMerge: Boolean = false): LogicalPlan = {
+
+    val scanBuilder = mergeBuilder.asScanBuilder
+    val scan = scanBuilder.build()
+    val scanRelation = DataSourceV2ScanRelation(table, scan, toOutputAttrs(scan.readSchema(), output))
+
+    scan match {
+      case filterable: SupportsFileFilter =>
+        if (performCountCheckForMerge) {
+          val affectedFilesAcc = new SetAccumulator[String]()
```
Comment (aokolnychyi) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: One thing I should mention here (not a blocker). The accumulator approach may take a substantial amount of driver memory. The problem happens when each executor accumulates a set of entries that covers a large portion of files to overwrite. For example, if you have to overwrite 10K files to perform a merge and each executor accumulates a set of let's say 5K and you have 200 executors, then you bring all 200 intermediate sets on the driver and then merge them into a single set. This may put a lot of pressure on the driver.

Comment (rdblue) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: This is a big motivation for implementing this a different way. I don't think that it is a good idea to use an accumulator. In addition to taking driver memory, accumulator state is passed in heartbeat messages and written to history files. We've had a problem with both in the past.

Comment (dilipbiswal) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: +1 Anton. I was thinking, we may think of providing a few alternate implementation of doing the cardinality check via options ?

Comment (dilipbiswal) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: @rdblue Oh.. oops .  sorry, didn't see your comment before putting mine.  I went ahead with accumulator based approach as delta oss also does the cardinality check the same way. Open to alternate suggestion on how we want to do the check.

Comment (rdblue) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: I'm okay moving forward with this to get something working, I guess. But I think we will need to replace it with something that does not use accumulators as soon as we can.

Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: I think using an accumulator name that starts with `internal.metrics` prevents tracking it in Spark UI but I am not sure about heartbeats and history files.

Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: I had one more idea in the original [comment](https://github.com/apache/iceberg/pull/1947#issuecomment-747450897) but I am not a fan of it either.

Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: Any alternatives, @dilipbiswal @rdblue?

Comment (rdblue) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: Yeah, I think we can add requirements for distribution and ordering to the MergeIntoExec node. That way we can guarantee that duplicates are on a single node and are clustered. Then we can do the row identifier check I suggested.

Comment (dilipbiswal) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: @aokolnychyi Yeah, i looked at option-2 but preferred option-1, as i think it would be more performant for many cases. Another idea i was thinking is, if it is possible for driver to broadcast a list of files to the executor and we only accumulate the index of the file instead of file name themselves. Not sure how expensive is to broadcast the files.. 
Intuitively it seems to me that we may not have a perfect solution that fits all the cases and was thinking if we could switch between implementations based on options.

Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: @rdblue, doing a join on the merge condition should guarantee that the same entries are co-located in one task so it should be enough to just add a local sort after the shuffle?

Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: @dilipbiswal, yeah, I agree option 2 is no good.

Comment (rdblue) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala:57]: @aokolnychyi, yeah, that's exactly the idea. And in most cases, the required local sort can be skipped because a sort-merge join would use one anyway.

The main problem is a broadcast join because either side of the join could be broadcasted. If the target table side is broadcasted instead of the update side, then there is no guarantee that rows are distributed by the merge condition because broadcast joins have no distribution requirement. So you would need to add both distribution and sort requirements. I did some looking around in Spark and I think that we can do that in a way that doesn't actually repartition when using a sort-merge join, but it looks like any broadcast join would get repartitioned.

There are 3 cases:
1. Sort-merge join is used: this works because the distribution/sort requirements are already satisfied
2. The target table side is large and the update side is broadcasted: redistribution and sort isn't needed, but we may not be able to avoid it.
3. The target table side is small and is broadcasted: this is the one that requires a distribution to co-locate rows

Here's an example:
```sql
MERGE INTO target t USING source s ON t.id = s.id
WHEN MATCHED UPDATE SET t.count = t.count + 1
WHEN NOT MATCHED INSERT (id, count) VALUES (s.id, 1)
```

Case 3 could happen when `target` has 2 rows and `source` has 10 rows. There is a good chance that `target` will be broadcasted to the partitions that hold `source`, and we don't have guarantees about how `source` is partitioned in Spark so it may have the same `id` in two partitions.

I think we would use `Hash(s.id, P)` and `SortOrder(s.id, ...)` for the distribution and sort requirements of `MergeIntoExec` (`P` is shuffle parallelism). Then for each case:
1. The output distribution for a sort-merge join will be `PartitioningCollection(Hash(s.id, P), Hash(t.id, P))` and the sort will be `SortOrder(s.id, ...)` with an annotation that `s.id` and `t.id` are equivalent, which should satisfy both requirements
2. The distribution will be an expansion of the streamed table's distribution, which is probably `UnspecifiedDistribution`. In that case, it looks like Spark can't add the distribution for the broadcast side so this would result in a repartition and sort after the join. This would actually result in the "partially clustered distribution" I've been talking about for joins.
3. The distribution and sort should be applied, but should also be reasonably cheap because the target rows are small

I think it should be okay that case 2 gets repartitioned and sorted, but it is not ideal. If only we could guarantee that it won't be broadcasted.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala [hunk 4]
```diff
@@ -41,30 +45,33 @@ import org.apache.spark.sql.catalyst.plans.logical.ReplaceData
 import org.apache.spark.sql.catalyst.plans.logical.UpdateAction
 import org.apache.spark.sql.catalyst.rules.Rule
 import org.apache.spark.sql.catalyst.utils.RewriteRowLevelOperationHelper
+import org.apache.spark.sql.connector.catalog.Table
 import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
 import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation
-import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.functions.monotonically_increasing_id
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala:51]: It doesn't look like this is used anymore.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala [hunk 5]
```diff
@@ -106,5 +113,18 @@ case class RewriteMergeInto(conf: SQLConf) extends Rule[LogicalPlan] with Rewrit
...
+    def hasUnconditionalDelete(action: Option[MergeAction]): Boolean = {
+      action match {
+        case Some(DeleteAction(None)) => true
+        case _ => false
+      }
+    }
+    val propertyVal = PropertyUtil.propertyAsBoolean(table.properties(),
+      MERGE_WRITE_CARDINALITY_CHECK, MERGE_WRITE_CARDINALITY_CHECK_DEFAULT)
+    val hasOneUnconditionalDelete = actions.size == 1 && hasUnconditionalDelete(actions.headOption)
+    !hasOneUnconditionalDelete && propertyVal
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala:127]: If I understand this logic correctly, the idea is that if the action does not depend on the source row at all, then don't perform the cardinality check?

That makes sense to me, but I would move the logic out of this method that returns whether the check is enabled. I consider the check enabled if the table property is enabled. The logic to skip the check based on the structure of the MERGE command should be separate so that we can maintain it separately.

So I'd split this into `isCardinalityCheckEnabled(table)` and `isCardinalityCheckNeeded(actions)`, then combine those in the rule: `val runCheck = isCardinalityCheckEnabled(table) && isCardinalityCheckNeeded(actions)`

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala [hunk 6]
```diff
@@ -67,8 +83,17 @@ trait RewriteRowLevelOperationHelper extends PredicateHelper with Logging {
 
     scan match {
       case filterable: SupportsFileFilter =>
-        val matchingFilePlan = buildFileFilterPlan(scanRelation.output, matchingRowsPlanBuilder(scanRelation))
-        DynamicFileFilter(scanRelation, matchingFilePlan, filterable)
+        if (performCountCheckForMerge) {
+          val affectedFilesAcc = new SetAccumulator[String]()
+          spark.sparkContext.register(affectedFilesAcc, AFFECTED_FILES_ACC_NAME)
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala:88]: What if this accumulator has already been used? Will you get the results from the previous accumulator if you run two MERGE INTO statements on the same table?

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala [hunk 7]
```diff
@@ -108,6 +133,23 @@ trait RewriteRowLevelOperationHelper extends PredicateHelper with Logging {
     Project(Seq(findOutputAttr(agg.output, FILE_NAME_COL)), agg)
   }
 
+  private def buildPlanWithFileAccumulator( fileAccum: SetAccumulator[String],
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala:136]: Style nit: the space between `(` and `fileAccum` is unnecessary.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala [hunk 8]
```diff
@@ -108,6 +133,23 @@ trait RewriteRowLevelOperationHelper extends PredicateHelper with Logging {
...
+                                            prunedTargetPlan: LogicalPlan): LogicalPlan = {
+    val fileAttr = findOutputAttr(prunedTargetPlan.output, FILE_NAME_COL)
+    val rowPosAttr = findOutputAttr(prunedTargetPlan.output, ROW_POS_COL)
+    val accumulatorExpr = Alias(AccumulateFiles(fileAccum, fileAttr), AFFECTED_FILES_ACC_ALIAS_NAME )()
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala:140]: Style nit: the space between `AFFECTED_FILES_ACC_ALIAS_NAME` and `)` is unnecessary.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/expressions/AccumulateFiles.scala [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.spark.sql.types.{DataType, IntegerType}
+
+case class AccumulateFiles(
+    filesAccumulator: SetAccumulator[String],
+    child: Expression) extends UnaryExpression with CodegenFallback {
+  override def dataType: DataType = IntegerType
+
+  override def nullable: Boolean = true
+  override def prettyName: String = "AccumulateFiles"
+  override lazy val deterministic = false
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/expressions/AccumulateFiles.scala:34]: Why isn't this deterministic? Are you setting this to avoid pushdown or other optimization?

File: spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DynamicFileFilterExec.scala [hunk 10]
```diff
@@ -47,13 +49,45 @@ case class DynamicFileFilterExec(
...
+    filesAccumulator: SetAccumulator[String],
+    @transient targetTableName: String)
+  extends DynamicFileFilterExecBase(scanExec, fileFilterExec, filterable)  {
+
+  override protected def doPrepare(): Unit = {
+    val rows = fileFilterExec.executeCollect()
+    if (rows.size > 0) {
+      val msg =
+        s"""
+           |The same row of target table `$targetTableName` was identified more than
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DynamicFileFilterExec.scala:85]: Nit: this error message starts with a newline. Can you start with `s"""The ...` instead?

File: spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DynamicFileFilterExec.scala [hunk 11]
```diff
@@ -47,13 +49,45 @@ case class DynamicFileFilterExec(
   override protected def doExecute(): RDD[InternalRow] = scanExec.execute()
   override protected def doExecuteColumnar(): RDD[ColumnarBatch] = scanExec.executeColumnar()
 
+  protected def doPrepare(): Unit
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DynamicFileFilterExec.scala:52]: Is this needed? I thought this was defined by `SparkPlan`

File: spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala [hunk 12]
```diff
@@ -30,14 +30,7 @@ import org.apache.spark.sql.catalyst.expressions.Expression
...
 import org.apache.spark.sql.catalyst.planning.PhysicalOperation
-import org.apache.spark.sql.catalyst.plans.logical.AddPartitionField
-import org.apache.spark.sql.catalyst.plans.logical.Call
-import org.apache.spark.sql.catalyst.plans.logical.DropPartitionField
-import org.apache.spark.sql.catalyst.plans.logical.DynamicFileFilter
-import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
-import org.apache.spark.sql.catalyst.plans.logical.MergeInto
-import org.apache.spark.sql.catalyst.plans.logical.ReplaceData
-import org.apache.spark.sql.catalyst.plans.logical.SetWriteOrder
+import org.apache.spark.sql.catalyst.plans.logical.{AddPartitionField, Call, DropPartitionField, DynamicFileFilter, DynamicFileFilterWithCountCheck, LogicalPlan, MergeInto, ReplaceData, SetWriteOrder}
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala:33]: Can you revert this? There are fewer commit conflicts if each import has its own line.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/expressions/AccumulateFiles.scala [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
+import org.apache.spark.sql.catalyst.utils.SetAccumulator
+import org.apache.spark.sql.types.{DataType, IntegerType}
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/expressions/AccumulateFiles.scala:25]: Could you move imports to separate lines?

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java [hunk 14]
```diff
@@ -232,8 +235,65 @@ public void testSourceFromSetOps() throws NoSuchTableException {
     sql(sqlText, targetName);
     sql("SELECT * FROM %s ORDER BY id, dep", targetName);
     assertEquals("Should have expected rows",
-            ImmutableList.of(row(1, "emp-id-1"), row(2, "emp-id-2")),
-            sql("SELECT * FROM %s ORDER BY id ASC NULLS LAST", targetName));
+           ImmutableList.of(row(1, "emp-id-1"), row(2, "emp-id-2")),
+           sql("SELECT * FROM %s ORDER BY id ASC NULLS LAST", targetName));
```
Comment (rdblue) [inline spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java:239]: Can you revert the whitespace change in this method? I don't think it actually changed.

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java [hunk 15]
```diff
@@ -232,8 +235,65 @@ public void testSourceFromSetOps() throws NoSuchTableException {
...
+  }
+
+  @Test
+  public void testAmbiguousRowInTarget() throws NoSuchTableException {
```
Comment (rdblue) [inline spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java:243]: The problem isn't the row in target, it is that two source rows match. Could you update the name to `testMultipleUpdatesForTargetRow`?

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java [hunk 16]
```diff
@@ -232,8 +235,65 @@ public void testSourceFromSetOps() throws NoSuchTableException {
...
+
+    String tabName = catalogName + "." + "default.target";
+    String errorMsg = "The same row of target table `" + tabName + "` was identified more than\n" +
+            " once for an update, delete or insert operation of the MERGE statement.";
+    AssertHelpers.assertThrows("Should complain ambiguous row in target",
+           SparkException.class, errorMsg, () -> sql(sqlText, targetName, sourceName));
+  }
+
+  @Test
+  public void testDisableAmbiguousRowInTargetCheck() throws NoSuchTableException {
```
Comment (rdblue) [inline spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java:261]: How about `testIgnoreMultipleUpdatesForTargetRow` or `testDisableCardinalityCheck`?

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java [hunk 17]
```diff
@@ -232,8 +235,65 @@ public void testSourceFromSetOps() throws NoSuchTableException {
...
+            "ON target.id = source.id " +
+            "WHEN MATCHED AND target.id = 1 THEN UPDATE SET * " +
+            "WHEN MATCHED AND target.id = 6 THEN DELETE " +
+            "WHEN NOT MATCHED AND source.id = 2 THEN INSERT * ";
+
+    String tabName = catalogName + "." + "default.target";
+    String errorMsg = "The same row of target table `" + tabName + "` was identified more than\n" +
+            " once for an update, delete or insert operation of the MERGE statement.";
+    AssertHelpers.assertThrows("Should complain ambiguous row in target",
+           SparkException.class, errorMsg, () -> sql(sqlText, targetName, sourceName));
```
Comment (rdblue) [inline spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java:257]: Can you also assert that the contents of the table are unchanged?

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java [hunk 18]
```diff
@@ -232,8 +235,65 @@ public void testSourceFromSetOps() throws NoSuchTableException {
...
+  }
+
+  @Test
+  public void testDisableAmbiguousRowInTargetCheck() throws NoSuchTableException {
+    append(targetName, new Employee(1, "emp-id-one"), new Employee(6, "emp-id-6"));
+    append(sourceName, new Employee(1, "emp-id-1"), new Employee(1, "emp-id-1"),
+           new Employee(2, "emp-id-2"), new Employee(6, "emp-id-6"));
```
Comment (rdblue) [inline spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java:264]: I think it would be better if the query here is the same one that fails in the test above. That way identical queries with identical data have different behavior based on the table property.

Comment (rdblue) [reply spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java:264]: Another reason to do this is that the merge always deletes, which looks like it may hit the case where an unconditional delete is allowed.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala (hunk 6 above)
Comment (rdblue) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala:88]: If it already exists, we probably need to call `reset` on it.

Comment (rdblue) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala:88]: Looks like the name is only used here. Maybe we should use a random name and each accumulator will be independent?

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/expressions/AccumulateFiles.scala (hunk 9 above)
Comment (dilipbiswal) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/expressions/AccumulateFiles.scala:34]: @rdblue Right Ryan. It's probably not needed. Since this is updating an accumulator and in a way has a state to it, i felt its better to make is non-deterministic.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DynamicFileFilterExec.scala (hunk 11 above)
Comment (dilipbiswal) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DynamicFileFilterExec.scala:52]: Not needed Ryan. removing it.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala (hunk 6 above)
Comment (dilipbiswal) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala:109]: @rdblue I debugged into the register code. We seem to be okay here. The registration actually happens by the address of the object and not by the name (to the best of my understanding).

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala [hunk 19]
```diff
@@ -47,12 +56,16 @@ import org.apache.spark.sql.sources
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.sql.util.CaseInsensitiveStringMap
 
+
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala:59]: Nit: unnecessary whitespace change

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala (hunk 6 above)
Comment (rdblue) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/RewriteRowLevelOperationHelper.scala:109]: Thanks for looking into it. Sounds good to me.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks good to me! I'll merge this when the JDK 8 tests pass.
